WalletConnect/WCInteractor.swift

```diff
@@ -161,58 +165,65 @@ extension WCInteractor {
-                guard let param = request.params.first else {
-                    throw WCError.badJSONRPCRequest
-                }
-                if param.approved == false {
-                    disconnect()
+    private func handleEvent(_ event: WCEvent, topic: String, decrypted: Data) throws {
+        switch event {
+        // topic == session.topic
+        case .sessionRequest:
+            let request: JSONRPCRequest<[WCSessionRequestParam]> = try event.decode(decrypted)
+            guard let params = request.params.first else {
+                throw WCError.badJSONRPCRequest
+            }
+            handshakeId = request.id
+            peerId = params.peerId
+            peerMeta = params.peerMeta
+            onSessionRequest?(request.id, params.peerMeta)
+        // topic == clientId
+        case .ethSign, .ethPersonalSign, .ethSignTypeData:
+            let request: JSONRPCRequest<[String]> = try event.decode(decrypted)
+            guard request.params.count > 1 else { throw WCError.badJSONRPCRequest }
+            let payload: WCEthereumSignPayload = {
+                if event == .ethSign {
+                    return .sign(data: Data(hex: request.params[1]), raw: request.params)
+                } else if event == .ethPersonalSign {
+                    return .personalSign(data: Data(hex: request.params[0]), raw: request.params)
+                } else if event == .ethSignTypeData {
+                    let data = request.params[0].data(using: .utf8) ?? Data()
+                    return .signTypeData(data: data, raw: request.params)
+                } else {
+                    fatalError()
                 }
-            default:
-                break
+            }()
+            onEthSign?(request.id, payload)
+        case .ethSendTransaction, .ethSignTransaction:
+            let request: JSONRPCRequest<[WCEthereumTransaction]> = try event.decode(decrypted)
+            guard request.params.count > 0 else { throw WCError.badJSONRPCRequest }
+            onEthTransaction?(request.id, event, request.params[0])
+        case .bnbSign:
+            if let request: JSONRPCRequest<[WCBinanceTradeOrder]> = try? event.decode(decrypted) {
+                onBnbSign?(request.id, request.params[0])
+            } else if let request: JSONRPCRequest<[WCBinanceCancelOrder]> = try? event.decode(decrypted) {
+                onBnbSign?(request.id, request.params[0])
+            } else if let request: JSONRPCRequest<[WCBinanceTransferOrder]> = try? event.decode(decrypted) {
+                onBnbSign?(request.id, request.params[0])
             }
-        } catch let error {
-            print("==> handleEvent error: \(error.localizedDescription)")
+            break
+        case .bnbTransactionConfirm:
+            let request: JSONRPCRequest<[WCBinanceTxConfirmParam]> = try event.decode(decrypted)
+            guard request.params.count > 0 else {
```

Comment:
What is the difference of validating either `count` or `first` object property? Could we make validation more generalized and move it to `decode` function?

Reply:
it depends on each event, some require two parameters, and some require only one. `decode` is generic method and doesn't know about types